### PR TITLE
Transpile TEAL code to substitute the TMPL placeholders

### LINKED ISSUE #343: Support PyTEAL Tmpl
It would be nice to support the `Tmpl` expression in pyteal and have a `parseArgs` function that can replace these variables. Each time I want to export a smart contract to a different project I need to manuallly edit all my smart contracts after testing to use the `Tmpl` expression.

### PR DESCRIPTION
**Pull request by amityadav0**
Closes #343

### CODE REVIEW

File: packages/algob/src/lib/compile.ts [hunk 1]
```diff
@@ -161,6 +178,18 @@ export class PyCompileOp {
...
+   * Replaces parameters with the values passed
+   * @param program Teal program
+   * @param replaceParams params that needs to be replaced in program
+   */
+  replaceTempValues (program: string, replaceParams: ReplaceParams): string {
+    for (const param in replaceParams) {
+      program = program.split(param).join(replaceParams[param]);
+    }
+    return program;
+  }
```
Comment (ratik21) [inline packages/algob/src/lib/compile.ts:191]: Shall we add test(s) ?

**Review by ratik21 — COMMENTED** [review]
Tests are missing. Can we add tests for template substitution (`TMPL_`) when compiling pyTEAL ?

File: packages/algob/src/types.ts [hunk 2]
```diff
@@ -612,6 +612,10 @@ export interface SCParams {
   [key: string]: string | bigint
 }
 
+export interface ReplaceParams {
```
[the remark this answers is not in the document]
Comment (amityadav0) [reply packages/algob/src/types.ts:615]: but it makes sense when we use it in code, for example `AccountAddress` it is only a string in context of type.

File: packages/algob/src/lib/compile.ts [hunk 3]
```diff
@@ -161,6 +178,18 @@ export class PyCompileOp {
     return pyCompiled;
   }
 
+  /**
+   * Replaces parameters with the values passed
+   * @param program Teal program
```
[the remark this answers is not in the document]
Comment (amityadav0) [reply packages/algob/src/lib/compile.ts:183]: TEAL code in string, i will update it to be more clear

File: packages/algob/src/lib/compile.ts [hunk 4]
```diff
@@ -161,6 +178,18 @@ export class PyCompileOp {
     return pyCompiled;
   }
 
+  /**
+   * Replaces parameters with the values passed
```
[the remark this answers is not in the document]
Comment (amityadav0) [reply packages/algob/src/lib/compile.ts:182]: could you please suggest ?

File: packages/algob/src/lib/compile.ts [hunk 5]
```diff
@@ -130,11 +130,28 @@ export class PyCompileOp {
...
-    console.log("PyTEAL template parameters:", scTmplParams);
-    let param: string | undefined = YAML.stringify(scTmplParams);
-    if (scTmplParams === undefined) { param = undefined; }
+    let param: string | undefined;
+    const replaceParams: ReplaceParams = {};
+    if (scTmplParams === undefined) {
+      param = undefined;
+    } else {
+      const tmp: SCParams = {};
+      for (const i in scTmplParams) {
```
[the remark this answers is not in the document]
Comment (amityadav0) [reply packages/algob/src/lib/compile.ts:139]: yes

File: packages/algob/src/lib/compile.ts [hunk 6]
```diff
@@ -130,11 +130,28 @@ export class PyCompileOp {
...
+        if (i.startsWith("TMPL_") || i.startsWith("tmpl_")) {
+          replaceParams[i] = scTmplParams[i].toString();
+        } else {
+          tmp[i] = scTmplParams[i];
+        }
+      }
+      console.log("PyTEAL template parameters:", tmp);
+      param = YAML.stringify(tmp);
+    }
+    console.log("TEAL replacement parameters:", replaceParams);
```
[the remark this answers is not in the document]
Comment (amityadav0) [reply packages/algob/src/lib/compile.ts:149]: which part ? parsing of scTmplParams ?

File: packages/algob/src/lib/compile.ts (hunk 1 above)
Comment (amityadav0) [reply packages/algob/src/lib/compile.ts:191]: i have updated the example to test this functionality

Comment (amityadav0) [reply packages/algob/src/lib/compile.ts:191]: let's add mocks also

File: packages/algob/src/lib/compile.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ratik21) [reply packages/algob/src/lib/compile.ts]: yes

File: packages/algob/src/types.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ratik21) [reply packages/algob/src/types.ts:615]: Actually we introduced `AccountAddress` for map type in runtime, eg. `new Map<number, AccountAddress>` is more intuitive than `new Map<number, string>`, as the latter looks more ambigous. But when used in functions i think `StringMap` is intuitive enough. Your call.

Anyway, please add a doc string `eg. { "ARG_TO": <address> }` to this interface then

File: docs/guide/py-teal.md [hunk 7]
```diff
@@ -58,29 +58,56 @@ sys.path.insert(0, path) # replace path with path to algobpy in your project
...
+- You can also use this feature with external support parameters: Ex:
+   ```js
+    scInitParam = {
+      TMPL_TO: "ADDR",
+      ARG_AMT: 700000,
+      ARG_CLS: masterAccount.addr
+    }
+    await deployer.loadLogic("dynamic-fee.py", scInitParam);
+   ```
+  Keys starting with `TMPL_` or `tmpl_` will be used with TMPL function and other keys will be used as mentioned in `External Parameters Support` section.
```
Comment (ratik21) [inline docs/guide/py-teal.md:108]: Please add this as a **NOTE:**

File: CHANGELOG.md [hunk 8]
```diff
@@ -18,6 +18,7 @@
 + Exported `@algorand-builder/algob`, `@algorand-builder/runtime` error types and make it accessible for API documentation.
 + Added `debugStack` option in `runtime.executeTx()` to print stack (upto depth = debugStack) after each opcode execution.
 + Tealv3 support in `@algo-builder/runtime`: `getBit`, `setBit`
++ Transpile TEAL code to substitute the TMPL placeholders
```
Comment (ratik21) [inline CHANGELOG.md:21]: Should we add this in **##Breaking** section? What happens if user is already using `TMPL_` prefixes in their `.py` contracts (like we were using before)?

**Review by ratik21 — APPROVED** [review]

File: packages/algob/src/types.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (amityadav0) [reply packages/algob/src/types.ts:615]: i think what we have is more intuitive, what do you mean ? "ARG_TO" is different then "TEMP_TO"

[the remark this answers is not in the document]
Comment (amityadav0) [reply packages/algob/src/types.ts:615]: also it is not necessary to have address as value of map, what kind of docs ? this is a map which has key as string and value as string which we can replace

[the remark this answers is not in the document]
Comment (amityadav0) [reply packages/algob/src/types.ts:615]: we also have `AccountMap` or `NetworksMap`. we use these interfaces for better understanding

File: CHANGELOG.md (hunk 8 above)
Comment (amityadav0) [reply CHANGELOG.md:21]: not sure if it's API breaking changes

Comment (amityadav0) [reply CHANGELOG.md:21]: these are breaking though

File: packages/algob/src/types.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ratik21) [reply packages/algob/src/types.ts:615]: > i think what we have is more intuitive, what do you mean ? "ARG_TO" is different then "TEMP_TO"
> this is a map which has key as string and value as string which we can replace

Exactly, that's why when i looked at it, i immediately thought it should be `StringMap`. Had to roam around the code to check why ReplaceParams was named as is. Suggested to add an `eg. { "ARG_TO": <address> }` doc string for better understanding.

`AccountMap`, `AnyMap`..etc seem clear enough.

**Pull request merged**